### PR TITLE
Updates sheets api version to v4

### LINKED ISSUE #367: Use Google Sheets API v4
#### Software and hardware versions 
Collect v1.4.14

#### Problem description
Collect currently uses an older API to interface with Google Sheets. This has probably contributed to minification issues and is a barrier to updating other libraries.

#### Other information 
The first question to answer is whether it will be possible to upgrade this library in isolation or whether it will also require updating other libraries such as Google Play Services. As much as possible, this upgrade should be isolated from others.

I looked at [Google's quick start guide](https://developers.google.com/sheets/api/quickstart/android) and their example requires several other updated libraries. I don't know whether this is for the rest of their example or whether they are truly needed to use `com.google.apis:google-api-services-sheets:v4-rev108-1.22.0`. I tried only adding
```
    compile('com.google.apis:google-api-services-sheets:v4-rev108-1.22.0') {
        exclude group: 'org.apache.httpcomponents'
    }
```
to build.gradle and removing `google-api-client-1.20.0.jar` which gets replaced by 1.22. That does compile although I didn't look to see if the API is useable. That would be the first thing to verify.

The [migration guide](https://developers.google.com/sheets/api/guides/migration) from v3 to v4 will probably be helpful.

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Fixes #367 
@grzesiek2010 @lognaturel I have done manual testing of uploading data to sheets. Please do check once again too as many libraries had to be upgraded. Also give reviews regarding the code structure and how it can be improved.
I used Geo Tagger v2 form for testing  

(P.S. I have not changed the code for uploading the photos, so it's not working. The whole picasa library needs to updated as well. Some of the protocols used are not supported anymore)
Thanks :smiley:

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@shobhitagarwal1612 I tried but it doesn't work when I try to send data:
screenshot_2017-03-15-10-21-27

any idea why?

**Comment by shobhitagarwal1612** [comment]
@grzesiek2010 Can you provide the sample of the form that you are using and the url of the sheets?

**Comment by grzesiek2010** [comment]
@shobhitagarwal1612  I'm sure it doesn't matter. I used simple form that contains only StringWidget and NumberWIdget (but you can try one of existed forms - as I remember I tested "Cascading Select Form" as well). I create new sheet everytime when I want to test this functionality so you can do the same you don't need mine. 

Do you have a problem with reproduction? Does it work for you?

**Comment by shobhitagarwal1612** [comment]
Yes I can't reproduce the problem. I used the Geo tagger v2 form available from GET FORMS section and it worked fine for me

**Comment by shobhitagarwal1612** [comment]
The error says that the range can't be parsed. It means that the first row of the Google sheet can't be parsed. Can you please try creating a new sheet and using that URL for testing?

**Comment by grzesiek2010** [comment]
I tried "Geo tagger v2" and it's the same. Did you try to send a form to completely empty sheet?

**Comment by shobhitagarwal1612** [comment]
Yes I found the problem. I used an old spreadsheet which already had empty cells. When I created a new sheet, I got the same error. Thanks for your time. I'll look into it now

**Comment by grzesiek2010** [comment]
@shobhitagarwal1612 it doesn't work like before. Please try to send one form then for example insert new column manually (in your sheet) after first column and try to send next form. 

I mean all values should be filled right.

**Comment by shobhitagarwal1612** [comment]
https://docs.google.com/spreadsheets/d/1SXAfR3ip6hpTm8xU492gfB_cgsE1xlscHAk1ozXs2DM/edit#gid=0
I uploaded the `cascading select form` to this spreadsheet (created a new one). The form successfully uploaded! Is it still showing the `Bad request` error?

**Comment by shobhitagarwal1612** [comment]
>it doesn't work like before

@grzesiek2010 Is there any unusual behaviour?

**Comment by grzesiek2010** [comment]
> https://docs.google.com/spreadsheets/d/1SXAfR3ip6hpTm8xU492gfB_cgsE1xlscHAk1ozXs2DM/edit#gid=0

So now insert an empty column after column A-form-county and try to send next form.

**Comment by shobhitagarwal1612** [comment]
I did as you mentioned and the new row got inserted. Is it unusual behaviour? Do we need to throw an error if the sheet contains extra columns?

**Comment by grzesiek2010** [comment]
@shobhitagarwal1612 let's talk on Slack :)

File: collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java [hunk 1]
```diff
@@ -147,27 +166,201 @@ private void runTask() {
...
+                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
+        return (networkInfo != null && networkInfo.isConnected());
+    }
+
+    /**
+     * Attempt to resolve a missing, out-of-date, invalid or disabled Google
+     * Play Services installation via a user dialog, if possible.
+     */
+    private void acquireGooglePlayServices() {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java]: This and the next method should probably go in `PlayServicesUtil`. And that would get us most of the way to #490 I believe.

**Comment by lognaturel** [comment]
It's looking really close! Nice work. 

Can you please add attributions in the comments for some of the methods you got from Google samples or other sources? That will help us find them if we ever have questions about that code in the future and also satisfy [attribution requirements](https://developers.google.com/terms/site-policies#attribution).

**Comment by shobhitagarwal1612** [comment]
@lognaturel Thanks :smiley:

**Comment by lognaturel** [comment]
With a setup that worked previously, I'm now getting the following issue when uploading a form:
```
W/Auth    ( 5659): [GetToken] GetToken failed with status code: NeedPermission
W/Auth    (10587): [GoogleAuthUtil] GoogleAuthUtil
W/System.err(10587): com.google.android.gms.auth.UserRecoverableAuthException: NeedPermission
W/System.err(10587): 	at com.google.android.gms.auth.zze$1.zzat(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.zze$1.zzau(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.zze.zza(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.zze.zzc(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.zze.getToken(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.zze.getToken(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.zze.getToken(Unknown Source)
W/System.err(10587): 	at com.google.android.gms.auth.GoogleAuthUtil.getToken(Unknown Source)
W/System.err(10587): 	at com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential.getToken(GoogleAccountCredential.java:269)
W/System.err(10587): 	at org.odk.collect.android.activities.GoogleSheetsUploaderActivity$GoogleSheetsInstanceUploaderTask.doInBackground(GoogleSheetsUploaderActivity.java:552)
W/System.err(10587): 	at org.odk.collect.android.activities.GoogleSheetsUploaderActivity$GoogleSheetsInstanceUploaderTask.doInBackground(GoogleSheetsUploaderActivity.java:525)
W/System.err(10587): 	at android.os.AsyncTask$2.call(AsyncTask.java:292)
W/System.err(10587): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
W/System.err(10587): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
W/System.err(10587): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
W/System.err(10587): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
W/System.err(10587): 	at java.lang.Thread.run(Thread.java:818)
```

It looks like [this](http://stackoverflow.com/a/14386248/137744) might address it!

**Comment by shobhitagarwal1612** [comment]
Android studio collapsed the catch blocks into one and removed this catch statement. I'll restore it back

**Comment by shobhitagarwal1612** [comment]
@lognaturel Is the issue still occurring?

**Comment by lognaturel** [comment]
@shobhitagarwal1612 Thanks, that fixed it! 

The last problem I can find is that it looks like questions that are left blank aren't accounted for in the results spreadsheet.

In the screenshot below, notice that for the 4th row, the UUID is actually in the location column. That's because I left the location and the picture blank, I believe.
screen shot 2017-03-15 at 10 09 33 pm

I'm using the demo form from [here](https://nafundi.com/blog/posts/serverless-odk-deployments-with-google-apps/).

**Comment by shobhitagarwal1612** [comment]
@lognaturel Thanks for your time

screenshot
Fixed the issue

**Comment by shobhitagarwal1612** [comment]
@lognaturel As you might have noticed, I explicitly checked permissions for `Accounts` again. This was done for devices running on API level 23 and above. They explicitly require permissions to be reasked dynamically during run time. This needs to be done at other places as well such as when accessing Google Maps API.

**Comment by lognaturel** [comment]
@shobhitagarwal1612 Thanks so much for taking on this really important upgrade! 💯

Good call on the permissions, I think that's what @nikhilbalyan filed at #644 so hopefully we can get that addressed soon too!

**Comment by shobhitagarwal1612** [comment]
I noticed that we'll have to call for many permissions check. I created a class regarding the same problem for one of my app which handles all the different permissions checks. Maybe we could use the same to have maximum code reusability

**Comment by lognaturel** [comment]
@shobhitagarwal1612 Didn't notice this message! How about adding some more info on that idea in #644? If your app is open source, you can even link to the class.

**Pull request merged**